### ISSUE TITLE
reliable pod scheduling on multi-node cluster

### ISSUE THREAD

**Issue by pohly** [intel/pmem-csi#92]
The current approach is:
- when asked to create a volume, merely do a space check and pretend that the volume has been created
- when asked to make the volume available on a node, create it there

This approach can fail in various ways:
- single-node cluster: the first pod using the volume might get started at a time when other pods have already used up all available NVDIMM space (because the check for available space and actual allocation are separate)
- multi-node cluster: the first pod using the volume might get scheduled to a node with insufficient NVDIMM space instead of the node with enough free space
- multi-node cluster: after stopping the pod and restarting it (think database server update!) or after starting another app using the same PV, the new pod might get scheduled to a node that doesn't have the PV

**Comment by pohly** [comment]
I think we have now come to a conclusion on how to proceed. We need to support two different volume life cycles:
- ephemeral: the volume is created together with the pod once the pod has been assigned to a node *and* gets destroyed when the pod stops running
- persistent: volumes get created and destroyed independently of a pod

Despite the project's name (csi-pmem, for "*persistent* memory"), the ephemeral use case is actually more relevant for current applications. Unfortunately, it's implementation is limited by what Kubernetes currently supports. Let's look at the implementation.

First, persistent mode. That's easy, because it just follows the normal CSI model:
- a user creates a PVC, with or without `WaitForFirstConsumer`
- `CreateVolume` looks for a node with sufficient free space and creates the volume there
- it returns topology information that locks users of the volume to that node
- when a pod starts to run, `NodeStageVolume` formats the volume and `NodePublishVolume` bind-mounts it (splitting this into two steps is cleaner and allows sharing the same volume between pods on the same node - I have a test for this scenario pending upstream in https://github.com/kubernetes/kubernetes/pull/72002)
- `DeleteVolume` deletes the volume

What we need to document as warning to the user is that they need to consider hardware failures and scenarios where the node becomes unavailable, because then pods using the volume cannot run.

Now ephemeral mode. Conceptually, this should be handled by an inline volume specification that comes with the POD, because a PVC by definition has a different life cycle. Kubernetes only supports inline volumes for in-tree drivers, CSI support is getting discussed here: https://github.com/kubernetes/enhancements/pull/716. As I pointed out in that KEP PR, that raises the question of how the Kubernetes scheduler then can pick suitable nodes. Currently it assumes that there is unlimited space and a local ephemeral volume can always be created. We need to work with upstream to come to a proper solution.

In the meantime, we can continue using the current "fake PVC binding" approach. But we need to add volume deletion in `NodeUnstageVolume` and give better guidance to users on the limitations of the current implementation, which is:
* beware that topology information in the PV is only an approximation: available space gets checked at the time when the PV gets created, not when the pod actually runs. In a busy system it can happen that a volume cannot be created when a pod starts, which will cause pod creation to get stuck until space becomes available again on the chosen node.
* use `WaitForFirstConsumer` to minimize the time delta between PV creation and pod start
* the PV persists with out-dated topology information after the first pod ran that used it and (currently) must be deleted manually (to be tested)

@avalluri agreed?

**Comment by pohly** [comment]
I've not thought too much about how to choose between the two modes. Perhaps a parameter in the `StorageClass`?

**Comment by avalluri** [comment]
> I think we have now come to a conclusion on how to proceed. We need to support two different volume life cycles:
> 
> * ephemeral: the volume is created together with the pod once the pod has been assigned to a node _and_ gets destroyed when the pod stops running
> * persistent: volumes get created and destroyed independently of a pod
> 
> Despite the project's name (csi-pmem, for "_persistent_ memory"), the ephemeral use case is actually more relevant for current applications. Unfortunately, it's implementation is limited by what Kubernetes currently supports. Let's look at the implementation.
> 
> First, persistent mode. That's easy, because it just follows the normal CSI model:
> 
> * a user creates a PVC, with or without `WaitForFirstConsumer`
> * `CreateVolume` looks for a node with sufficient free space and creates the volume there
> * it returns topology information that locks users of the volume to that node

Here is my problem, locking PV at this phase and forcing Pod to run here might not be right solution in my opinion. As Pod might have other scheduler conditions(NodeAffinity, other resource requests) which might not fulfill by this Node, while at the sametime other Node(s) might satisfy those conditions and have enough PMEM capacity.

**Comment by pohly** [comment]
@avalluri so what are you suggesting? That pmem-csi should not support persistent volumes at all?

One cannot have persistent volumes that move between nodes. So either a user chooses to lock his pods onto one node (when using persistent mode) or he keeps his pods relocatable by giving up data persistency. This is the choice that users have to make, and then pmem-csi needs implement that as well as it can.

**Comment by avalluri** [comment]
> @avalluri so what are you suggesting? That pmem-csi should not support persistent volumes at all?
> 
My proposal is to lock PV on first usage/claim. That means,  Driver shall lock(by dynamically updating NodeAffinity) the PV to node where the pod scheduled. In initial(CreateVolume) phase we only give hints to scheduler where is the free space, instead of choosing node by driver.

**Comment by pohly** [comment]
I see several issues with this:
* It cannot be implemented using the current CSI API; we could do it by replacing kubernetes-csi/external-provisioner with our own which knows about changing topology, but then it still would only work on Kubernetes.
* Users who want persistent volumes cannot be sure that the space is really available in their cluster until they run a pod using it.
* Last but not least, I don't think that allocating space for a long time (which is what "persistent" mode implies) based on the current workload of a node (which is what the current scheduler would do) makes sense. I'm all for granting the user more control over the placement of persistent volumes, but it should be explicit with parameters in the PVC, not indirectly through the pod.

But feel free to bring up this third usage model in next weeks meeting to get feedback from the architects.

**Comment by okartau** [comment]
I can imagine a 3rd type of volume that could be useful in some scenarios, it's kind of "persistent with relaxed requirements". Lets vision a cluster of load-balancing similar workers that use local pmem-based volume for caching some data that's otherwise remotely accessed. Every worker handles different requests so their caches are no exactly same (and there is no requirement for caches to be same). That means persistent volume on every worker that survives reboot (thus speeding up operation after next boot) (means it is not ephemeral), and many of hard-to-achieve requirements we can skip (locking to node, where to start, keeping same content, replicate data etc). Does this scenario make any sense?
Is there already such type of volume which runs as similar instance on every node?

**Comment by pohly** [comment]
@okartau I think such a scenario makes perfect sense.

Only problem is, when I asked both the guys from Weave and Jim about an application that might benefit from such a cache, they couldn't (at least at that time) name one. But there must be something, because for example https://github.com/kubernetes/enhancements/pull/725 explicitly describes that use case (https://github.com/kubernetes/enhancements/blob/c4b378ed48fdfa6f78d6a07a2fff9393b9f65c68/keps/sig-storage/0034-20190124-local-persistent-volumes.md#caching: "Caching: ... there is a use case for persistent local storage to achieve data gravity for large caches. For large caches, if a pod restarts, rebuilding the cache can take a long time").

I'm not aware of a formally defined volume type for this. But we can map it onto the existing PVC/PV model:
- when the user creates a PVC, they specify (or let the system choose) a number of nodes on which a local volume is to be created
- `CreateVolume` then creates them
- `PV` has node affinity such that it only matches the nodes which have the local volume
- each time a pod uses the volume, they end up using the instance on their local node

The implication is that the app cannot rely on any data being in the volume, but if there is data, it can be sure that it was written by a previous incarnation of the app. That's fine for a cache.

**Comment by okartau** [comment]
Did recent csi-1.0 update cover concerns in this issue so that we can close it?

**Comment by avalluri** [comment]
> Did recent csi-1.0 update cover concerns in this issue so that we can close it?

Not fully covered, the csi-1.0 move make sure that:
- schedule the application(s) which claims the provisioned volume on right node(s) where the volume created
- In late binding mode, create the volume on Kubernetes selected node.

The major part still missing is "Kubernetes considering the node storage capacity while selecting the node for scheduling application in case of late binding".

**Comment by pohly** [comment]
Amarnath Valluri <notifications@github.com> writes:
> The major part still missing is "Kubernetes considering the node
> storage capacity while selecting the node for scheduling application
> in case of late binding".

@avalluri: can you try out what happens when Kubernetes selects a node
which then doesn't have the necessary capacity?

If I remember and understand
https://github.com/kubernetes-csi/external-provisioner/issues/221
correctly, the driver will then get an AccessibleTopology.Requisite with
multiple entries, i.e. multiple nodes. If the first entry doesn't have
enough capacity, it will respond by making the volume available on some
other node, right?

I expect pod startup to fail in this case because the pod has already
been scheduled onto a node on which the volume then cannot be mounted.

I think the driver's behavior is covered by the "if x<n, then the SP
SHALL choose x unique topologies from the list of requisite topologies"
case.

That we end up with an unschedulable pod is then a problem that needs to
be handled by Kubernetes and/or external-provisioner.

**Comment by avalluri** [comment]
> If I remember and understand [kubernetes-csi/external-provisioner#221](https://github.com/kubernetes-csi/external-provisioner/issues/221) correctly, the driver will then get an AccessibleTopology.Requisite with multiple entries, i.e. multiple nodes. If the first entry doesn't have enough capacity, it will respond by making the volume available on some other node, right?

Yes, the driver has no information if a node is  selected by Kuberentes or random selection of external-provisioner, so we just find and allocate the volume on a free node in the order of `AccessibleTopology.{Preferred,Requisite}`.

**Comment by avalluri** [comment]
> I think the driver's behavior is covered by the "if x<n, then the SP SHALL choose x unique topologies from the list of requisite topologies" case.

Yes, right.

> That we end up with an unschedulable pod is then a problem that needs to be handled by Kubernetes and/or external-provisioner.

As of my knowledge external-provisoner do not handle this case.

**Comment by avalluri** [comment]
> @avalluri: can you try out what happens when Kubernetes selects a node which then doesn't have the necessary capacity?

As of my observation Pod scheduling fails, I will try again and confirm.

**Comment by pohly** [comment]
Amarnath Valluri <notifications@github.com> writes:

>> @avalluri: can you try out what happens when Kubernetes selects a node which then doesn't have the necessary capacity?
>
> As of my observation Pod scheduling fails, I will try again and
> confirm.

While you do that, capture logs that show exactly what the topology in
the cluster is (CSINodeInfo) and what the sequence of events is (pod
scheduled, external-provisioner starts, gRPC calls and their
parameters). That then needs to go into
https://github.com/kubernetes-csi/external-provisioner/issues/221 to
show that something is amiss.

**Comment by avalluri** [comment]
@pohly I mimic the case of delayed PV binding where Pod scheduled on a node(host-2), but the volume created on different node/topology(host-3). As we expected scheduling failed, because of the volume locked to different topology/node which cannot be attached to Pods's node/topology :

```
$ kubectl describe pod my-csi-app
Events:
  Type     Reason              Age                From                     Message
  ----     ------              ----               ----                     -------
  Normal   Scheduled           62s                default-scheduler        Successfully assigned default/my-csi-app to host-2
  Warning  FailedAttachVolume  30s (x7 over 62s)  attachdetach-controller  AttachVolume.Attach failed for volume "pvc-bc6f6cb9-4721-11e9-a367-deadbeef0100" : rpc error: code = FailedPrecondition desc = Volume cannot be published on requested node host-2
```

Application Pod info:

```
$ kubectl get po -o wide
NAME                    READY   STATUS              RESTARTS   AGE     IP            NODE     NOMINATED NODE   READINESS GATES
my-csi-app              0/1     ContainerCreating   0          56s     <none>        host-2   <none>           <none>
```

create pv topology:
```
$ kubectl get pv -o json | jq '.items[] | .name, .spec.nodeAffinity'
"pvc-bc6f6cb9-4721-11e9-a367-deadbeef0100"
{
  "required": {
    "nodeSelectorTerms": [
      {
        "matchExpressions": [
          {
            "key": "pmem-csi.intel.com/node",
            "operator": "In",
            "values": [
              "host-3"
            ]
          }
        ]
      }
    ]
  }
}
```

But the question here is what is the expected behavior of a CSI driver in this case, 
a) if it should fail to create volume if there is no space on selected node, or
b) it just create the volume irrespective of node selection ?

If (a) then we need to negotiate this with external-provisioner, if not, then no change needed. But either case Kubernetes fails to schedule the Pod.

**Comment by avalluri** [comment]
The right solution in my opinion would be, to make the Kubernetes aware of storage capacity and consider that while scheduling the Pod.

**Comment by pohly** [comment]
Will be fixed in 0.7.0 through:
- scheduler extensions
- rescheduling of pods with late-binding volumes (a feature in external-provision 1.6.0, brought in by PR #627)

**Comment by pohly** [comment]
Resolved: scheduler extensions are merged and sidecars were updated.
